### PR TITLE
[#133] feat(netty): Rewrite protocol.

### PR DESCRIPTION
**Pull request by leixm**
### What changes were proposed in this pull request?
Rewrite protocol.

### Why are the changes needed?
For netty replace grpc.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?

UTs.

### CODE REVIEW

**Comment by jerqi** [comment]
We don't need to modify every protocol.  It's ok that we only modify the data panel. We could only modify ` GetLocalShuffleData`, `GetMemoryShuffleData` and `GetLocalShuffleIndex`.

**Comment by leixm** [comment]
> We don't need to modify every protocol. It's ok that we only modify the data panel. We could only modify ` GetLocalShuffleData`, `GetMemoryShuffleData` and `GetLocalShuffleIndex`.

It seems that it has no effect, we can implement this part of the protocol first, and then decide whether to modify all interfaces.

**Comment by leixm** [comment]
> We don't need to modify every protocol. It's ok that we only modify the data panel. We could only modify ` GetLocalShuffleData`, `GetMemoryShuffleData` and `GetLocalShuffleIndex`.

This will not block our progress much.

**Comment by jerqi** [comment]
> > We don't need to modify every protocol. It's ok that we only modify the data panel. We could only modify ` GetLocalShuffleData`, `GetMemoryShuffleData` and `GetLocalShuffleIndex`.
> 
> It seems that it has no effect, we can implement this part of the protocol first, and then decide whether to modify all interfaces.

This pr is a little big and it will make people confusing if some protocols aren't used but they are implemented.

**Comment by leixm** [comment]
@jerqi  @smallzhongfeng  PTAL.

File: common/src/main/java/org/apache/uniffle/common/ShuffleDataDistributionType.java [hunk 1]
```diff
@@ -17,10 +17,43 @@
...
+import org.apache.uniffle.common.netty.EncodeException;
+import org.apache.uniffle.common.netty.protocol.Encodable;
+
 /**
  * The type of shuffle data distribution of a single partition.
  */
-public enum ShuffleDataDistributionType {
-  NORMAL,
-  LOCAL_ORDER
+public enum ShuffleDataDistributionType implements Encodable {
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/ShuffleDataDistributionType.java:28]: Why do this class need implements `Encodable`?

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+  @Override
+  public int encodedLength() {
+    return super.encodedLength() + Integer.BYTES + data.readableBytes();
+  }
+
+  @Override
+  public void encode(ByteBuf buf) {
+    super.encode(buf);
+    ByteBufUtils.copyByteBuf(data, buf);
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java:50]: Could we reduce this copy? Do we need release this ByteBuf?

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/GetMemoryShuffleDataResponse.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  public int encodedLength() {
+    return super.encodedLength() + Encoders.encodeLengthOfBufferSegments(bufferSegments)
+        + Integer.BYTES + data.readableBytes();
+  }
+
+  @Override
+  public void encode(ByteBuf buf) {
+    super.encode(buf);
+    Encoders.encodeBufferSegments(bufferSegments, buf);
+    ByteBufUtils.copyByteBuf(data, buf);
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/netty/protocol/GetMemoryShuffleDataResponse.java:60]: ditto.

File: common/src/main/java/org/apache/uniffle/common/util/ByteBufUtils.java [hunk 4]
```diff
@@ -24,7 +24,10 @@
 public class ByteBufUtils {
 
   public static int encodedLength(String s) {
-    return 4 + s.getBytes(StandardCharsets.UTF_8).length;
+    if (s == null) {
```
Comment (smallzhongfeng) [inline common/src/main/java/org/apache/uniffle/common/util/ByteBufUtils.java:27]: We can use `StringUtils.isEmpty(s)`.

File: common/src/main/java/org/apache/uniffle/common/netty/EncodeException.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.uniffle.common.netty;
+
+public class EncodeException extends RuntimeException {
```
Comment (zuston) [inline common/src/main/java/org/apache/uniffle/common/netty/EncodeException.java:20]: extends `RssExcepption` ?

File: common/src/main/java/org/apache/uniffle/common/netty/DecodeException.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.uniffle.common.netty;
+
+public class DecodeException extends RuntimeException {
```
Comment (zuston) [inline common/src/main/java/org/apache/uniffle/common/netty/DecodeException.java:20]: extends RssExcepption ?

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleIndexResponse.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  @Override
+  public void encode(ByteBuf buf) {
+    super.encode(buf);
+    ByteBufUtils.copyByteBuf(indexData, buf);
+    buf.writeLong(fileLength);
+  }
+
+  public static GetLocalShuffleIndexResponse decode(ByteBuf byteBuf) {
```
Comment (zuston) [inline common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleIndexResponse.java:58]: If I understand correctly, this method is for client to decode into `GetLocalShuffleIndexResponse` from `byteBuf`.

Can we have a pluggable decoder/encoder to unify different responses and requests? 

The `decode` also should be as an interface method ?

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/GetMemoryShuffleDataRequest.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    buf.writeLong(timestamp);
+    buf.writeInt((int) expectedTaskIdsBitmap.serializedSizeInBytes());
+    try {
+      buf.writeBytes(RssUtils.serializeBitMap(expectedTaskIdsBitmap));
+    } catch (IOException ioException) {
+      throw new EncodeException("serializeBitMap failed while encode GetMemoryShuffleDataRequest!", ioException);
+    }
+  }
+
+  public static GetMemoryShuffleDataRequest decode(ByteBuf byteBuf) throws DecodeException {
```
Comment (zuston) [inline common/src/main/java/org/apache/uniffle/common/netty/protocol/GetMemoryShuffleDataRequest.java:79]: ditto

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/GetMemoryShuffleDataResponse.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  public static GetMemoryShuffleDataResponse decode(ByteBuf byteBuf) {
```
Comment (zuston) [inline common/src/main/java/org/apache/uniffle/common/netty/protocol/GetMemoryShuffleDataResponse.java:63]: ditto

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleIndexResponse.java (hunk 7 above)
Comment (leixm) [reply common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleIndexResponse.java:58]: Spark does the same, See https://github.com/apache/spark/blob/master/common/network-common/src/main/java/org/apache/spark/network/protocol/RpcResponse.java.

Comment (leixm) [reply common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleIndexResponse.java:58]: I prefer this way.

File: common/src/main/java/org/apache/uniffle/common/util/ByteBufUtils.java (hunk 4 above)
Comment (leixm) [reply common/src/main/java/org/apache/uniffle/common/util/ByteBufUtils.java:27]: `Null`and `an empty string` have different meanings, here we need to judge whether it is null.

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java (hunk 2 above)
Comment (leixm) [reply common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java:50]: Not yet, this is a todo
We discussed this issue in #826, in fact, the following two conditions need to be met to reduce this copy
1. Implement a MessageWithHeader
2. The above data is used as `body`

In this way, the data of data can be directly copied to`WritableByteChannel`, but there are some problems in the current implementation of point 2. For example, our SendShuffleData contains multiple blocks, and some modifications needs to be done, such as the metadata in ShuffleBlockInfo and binary data separated, In order to treat the binary data of all blocks as a `body`.

Comment (leixm) [reply common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java:50]: Sure, we need to release, i will fix.

**Comment by leixm** [comment]
> This pr title and summary is a bit short, could you elaborate a bit more on these?

Maybe this is more convenient for review.

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java (hunk 2 above)
Comment (leixm) [reply common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java:50]: `MessageWithHeader ` is introduced to support zero-copy, See https://github.com/apache/spark/blob/master/common/network-common/src/main/java/org/apache/spark/network/protocol/MessageWithHeader.java.

Comment (jerqi) [reply common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java:50]: cc @smallzhongfeng

**Comment by leixm** [comment]
@jerqi  @zuston  PTAL.

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    data.release();
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java:51]: ByteBuf release method don't guarantee the memory release? We will raise an issue to track this problems.
You can see `UnpooledDirectByteBuf(ByteBufAllocator alloc, ByteBuffer initialBuffer, int maxCapacity, boolean doFree, boolean slice)`. Some ByteBuf will be created with false doFree. These ByteBuf can't released by the method `release`. But we can skip this issue  in this pr. Let's check this issue in next prs.

**Comment by jerqi** [comment]
I prefer merging this pr as soon as we can, major comments have been addressed. If we have another comments, we can raise some follow-up prs.

**Review by jerqi — APPROVED** [review]
LGTM

**Comment by leixm** [comment]
Thanks for your review. @jerqi @advancedxy @zuston @smallzhongfeng

**Pull request merged**